src/middleware/open-graph/index.js

```diff
@@ -0,0 +1,180 @@
+			resolve(`data:${mime};base64,${Buffer.concat(logo).toString('base64')}`);
+		});
+
+		stream.on('error', reject);
+
+		stream.on('response', (res) => {
+			if (res.headers['content-length'] && res.headers['content-length'] > LOGO_MAX_SIZE) {
+				return stream.destroy(new Error('logo size limit exited'));
+			}
+
+			mime = res.headers['content-type'] ?? 'image/png';
+		});
+
+		stream.on('data', (chunk) => {
+			size += chunk.length;
+
+			if (size > LOGO_MAX_SIZE) {
+				return stream.destroy(new Error('logo size limit exited'));
+			}
+
+			logo.push(chunk);
+		});
+	});
+};
+
+const truncateString = (str, length) => {
+	if (str && str.length > length) {
+		return str.substr(0, length - 3) + '...';
+	}
+
+	return str;
+};
+
+const processDescription = (description) => {
+	let charsPerLine = 54;
+	let lineOffset = 209;
+	let lineHeight = 40;
+
+	let lines = wordwrap.hard(charsPerLine)(description).split('\n');
+
+	if (lines.length > 2) {
+		lines = lines.slice(0, 2);
+		let lastLine = lines.pop();
+
+		if (lastLine.length > charsPerLine) {
+			lastLine = lastLine.substr(0, charsPerLine - 3);
+		}
+
+		lines.push(lastLine + '...');
+	}
+
+	lines = lines.map((text, idx) => {
+		return { text, offset: lineOffset + lineHeight * idx };
+	});
+
+	return lines;
+};
+
+const processChart = (records, max, offsetX = 88, offsetY = 414) => {
+	let barHeight = 40;
+	let barPadding = 10;
+
+	return records.map((record, idx) => {
+		let h = max === 0 ? 0 : Math.floor((record.total / max) * barHeight);
+		let x = offsetX + barPadding * idx;
+		let y = offsetY + barHeight - h;
+
+		return { h, x, y };
+	});
+};
+
+const prepareMetadata = async (pkg) => {
+	let cacheKey = `algolia::${pkg}`;
+
+	if (cache.has(cacheKey)) {
+		return cache.get(cacheKey);
+	}
+
+	let meta = await npmIndex.getObject(pkg, [ 'name', 'description', 'version', 'owner' ]);
+
+	meta.owner.logo = await fetchLogo(meta.owner.avatar).catch(() => {});
+	meta.description = processDescription(meta.description);
+
+	cache.set(cacheKey, meta);
+
+	return meta;
+};
+
+const prepareStats = async (name) => {
+	let stats = await fetchStats(name);
+
+	let formatNum = num => num.toLocaleString().replaceAll(',', ' ');
+	let max = 0;
+
+	let records = Object.values(stats.dates).map((stats) => {
+		max = Math.max(max, stats.total);
+		return { total: stats.total };
+	});
+
+	return {
+		requests: {
+			total: stats.total,
+			totalFormatted: formatNum(stats.total),
+			chart: processChart(records, max),
+		},
+		bandwidth: {
+			total: stats.total,
+			totalFormatted: bytes(stats.total),
+			chart: processChart(records, max, 580),
+		},
+	};
+};
+
+const composeTemplate = async (name, scope = null) => {
+	let pkg = scope ? scope + '/' + name : name;
+	let [ metadata, stats ] = await Promise.all([ prepareMetadata(pkg), prepareStats(pkg) ]);
+
+	return {
+		name: truncateString(name, 24),
+		scope: truncateString(scope, 36),
+		description: metadata.description,
+		version: metadata.version,
```

Comment:
It seems the version/author texts currently don't have any truncate logic? That could be a problem too in some cases. But maybe wait with fixing this because we're considering some design changes too.

Reply:
Truncate logic that is simple and works even if we go with the dynamic positioning:

Cap the first field so that it ends at 542px from the left (simply remove letters from the end till it fits).

For the second field: assume start at 582px (which is a worst-case, assuming 542 + 40px padding), the right edge where the second chart ends, calculate fixed max-width that will be applied regardless of the actual position (i.e., the first field could be short, but we won't be changing the max-width for the second field based on that, to keep this simple).